### PR TITLE
Use a Table-backed RLMResults for allObjects

### LINKED ISSUE #1178: allObjects should return RLMArray backed by a Table
This will improve performance considerably, especially when processing notifications/updates for query-less accessors.

### PR DESCRIPTION
**Pull request by tgoyne**
Closes #1178.

@alazier @jpsim

### CODE REVIEW

File: Realm/Tests/ArrayPropertyTests.m [hunk 1]
```diff
@@ -334,4 +333,105 @@ - (void)testSortByMultipleColumns {
...
+    XCTAssertThrowsSpecificNamed([peopleInCompany removeObjectAtIndex:3], NSException, @"RLMException", @"Out of bounds");
+    XCTAssertNoThrow([peopleInCompany removeObjectAtIndex:1], @"Should delete link to employee");
+    [realm commitWriteTransaction];
+
+    XCTAssertEqual(peopleInCompany.count, 2U, @"link deleted when accessing via links");
+    EmployeeObject *test = peopleInCompany[0];
+    XCTAssertEqual(test.age, po1.age, @"Should be equal");
+    XCTAssertEqualObjects(test.name, po1.name, @"Should be equal");
+    XCTAssertEqual(test.hired, po1.hired, @"Should be equal");
+    //XCTAssertEqualObjects(test, po1, @"Should be equal"); //FIXME, should work. Asana : https://app.asana.com/0/861870036984/13123030433568
```
Comment (jpsim) [inline Realm/Tests/ArrayPropertyTests.m]: That Asana task was marked as completed 6 months ago. Does the test still fail? If so, we should open a new GitHub issue.

File: Realm/RLMObjectStore.mm [hunk 2]
```diff
@@ -431,14 +431,18 @@ void RLMDeleteAllObjectsFromRealm(RLMRealm *realm) {
...
-                                                                            realm:realm];
-    return results;
+
+    if (predicate) {
+        tightdb::Query query = objectSchema.table->where();
+        RLMUpdateQueryWithPredicate(&query, predicate, realm.schema, objectSchema);
+
+        // create and populate array
+        return [RLMResults resultsWithObjectClassName:objectClassName
+                                                query:std::make_unique<Query>(query)
```
Comment (alazier) [inline Realm/RLMObjectStore.mm]: You aligned the wrong :

File: Realm/RLMResults.mm [hunk 3]
```diff
@@ -295,25 +296,24 @@ - (id)objectAtIndexedSubscript:(NSUInteger)index {
     return [self objectAtIndex:index];
 }
 
--(id)minOfProperty:(NSString *)property {
-    RLMResultsValidate(self);
-
-    if (self.count == 0) {
+template<typename Table>
```
Comment (alazier) [inline Realm/RLMResults.mm]: Table for the template parameter name is confusing. TableType might be an improvement.

File: Realm/Tests/ArrayPropertyTests.m [hunk 4]
```diff
@@ -334,4 +333,105 @@ - (void)testSortByMultipleColumns {
     XCTAssertTrue(checkOrder(@[@"age", @"dogName"], @[@NO, @NO], @[b2, a2, b1, a1]));
 }
 
+- (void)testDeleteLinksAndObjectsInArray
+{
+    RLMRealm *realm = [RLMRealm defaultRealm];
+    [realm beginWriteTransaction];
+
+    EmployeeObject *po1 = [[EmployeeObject alloc] init];
```
Comment (alazier) [inline Realm/Tests/ArrayPropertyTests.m]: These 15 lines could be compacted to
```
[EmployeeObject createWithObject:@[@40, @"Joe", @YES]];
[EmployeeObject createWithObject:@[@ 30, @"John", @NO]];
[EmployeeObject createWithObject:@[@ 25, @"Jill", @YES]];
```

**Comment by alazier** [comment]
Do we have separate tests for any method which has different implementations for the two RLMResults variants?

Might also be nice to verify the subclass is what we expect in the tests.

Also, it would be great if there were a way to factor the test so that the same code could get applied to both variants. Not sure if this is possible though.

**Comment by jpsim** [comment]
While this looks good and mostly safe, I'm worried that our existing tests and these modifications to them don't cover quite enough.

**Comment by tgoyne** [comment]
Any suggestions for other things to test? RLMResults.mm now has 99% test coverage, and the only uncovered things are some error checking unrelated to the changes here. All of the checks which involve code-paths that vary between the two implementations are done on both.

**Comment by segiddins** [comment]
Also closes https://github.com/realm/realm-cocoa/issues/1160 ?

**Comment by tgoyne** [comment]
No, this has no effect once you sort or filter the results in any way.

**Comment by bmunkholm** [comment]
Nice to finally get that done :-)
Did you measure the performance improvements?

**Comment by jpsim** [comment]
@bmunkholm see https://github.com/realm/realm-cocoa/pull/1353/files#diff-1. Hard to tell if there were necessarily *improvements*, but there certainly weren't any regressions based on the performance tests we have.

**Comment by tgoyne** [comment]
This reduces the asymptotic complexity of many operations (O(N<sup>2</sup>) -> O(N) for enumerate-and-mutate, O(N) -> O(1) for just checking the count), so the speedup depends on the size of the data set. For the specific user report that lead to actually implementing this it's about 100x faster.

**Comment by alazier** [comment]
:+1:

**Pull request merged**